Sources/NIOCore/AsyncChannelAdapter.swift

```diff
@@ -0,0 +1,302 @@
+                        self._state = .bufferingWithoutPendingRead(buffer)
+                        return (result, nil)
+                    } else {
+                        self._state = .waitingForBuffer(buffer, continuation)
+                        return (nil, nil)
+                    }
+
+                case .bufferingWithPendingRead(var buffer, let promise):
+                    // No matter what happens here, we need to complete this promise.
+                    if let result = buffer.popFirst() {
+                        self._state = .bufferingWithoutPendingRead(buffer)
+                        return (result, promise)
+                    } else {
+                        self._state = .waitingForBuffer(buffer, continuation)
+                        return (nil, promise)
+                    }
+
+                case .waitingForBuffer:
+                    preconditionFailure("Should never be async nexting twice")
+                }
+            }
+
+            if let message = result.0 {
+                continuation.resume(returning: message)
+            }
+            if let promise = result.1 {
+                promise.succeed(())
+            }
+        }
+    }
+}
+
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+extension NIOAsyncChannelAdapterHandler {
+    @usableFromInline
+    enum InboundMessage {
+        case channelRead(InboundIn, EventLoopPromise<Void>?)
+        case eof
+    }
+}
+
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+extension NIOAsyncChannelAdapterHandler {
+    @usableFromInline
+    enum StreamState {
+        case bufferingWithoutPendingRead(CircularBuffer<InboundMessage>)
+        case bufferingWithPendingRead(CircularBuffer<InboundMessage>, EventLoopPromise<Void>)
+        case waitingForBuffer(CircularBuffer<InboundMessage>, CheckedContinuation<InboundMessage, Never>)
+    }
+}
+
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+public struct NIOInboundChannelStream<InboundIn>: @unchecked Sendable {
+    @usableFromInline let _handler: NIOAsyncChannelAdapterHandler<InboundIn>
+
+    @inlinable
+    public init(_ handler: NIOAsyncChannelAdapterHandler<InboundIn>) {
+        self._handler = handler
+    }
+}
+
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+extension NIOInboundChannelStream: AsyncSequence {
+    public typealias Element = InboundIn
+
+    public class AsyncIterator: AsyncIteratorProtocol {
+        @usableFromInline let _handler: NIOAsyncChannelAdapterHandler<InboundIn>
+
+        @inlinable
+        init(_ handler: NIOAsyncChannelAdapterHandler<InboundIn>) {
+            self._handler = handler
+        }
+
+        @inlinable public func next() async throws -> Element? {
+            switch await self._handler.loadEvent() {
+            case .channelRead(let message, let promise):
+                promise?.succeed(())
+                return message
+            case .eof:
+                return nil
+            }
+        }
```

Comment:
I know we are wrapping the calls to this to a `withTaskCancellationHandler(handler:operation:)` above but shouldn't we check `isCancelled` here as well. In the case users get it wrong?